docs/RFCS/20200811_non_blocking_txns.md

```diff
@@ -0,0 +1,1027 @@
+good, and, if the replication factor is 3, having 2 replicas in one region is
+also not good because it’s equivalent to having all of them in one region. What
+you have to do, I guess, is set a replication factor of at least 5 *for each
+index* and then configure two replicas in the reader region (that’s 3*5=15
+replicas for every index for a 3-region configuration; 10*5=50 replicas in a
+10-region configuration). In contrast, with this RFC's proposal, you'd have 2
+replicas per region (so 2 per region vs 5 per region). Our documentation doesn't
+talk about replication factors and fault tolerance. Probably for the better,
+because these issues are hard to reason about if you don't spend your days
+counting quorums. This point shows the problem with doing such replication
+schemes at the SQL level: we hide from the system the redundancy between the
+indexes (and between the replicas of different indexes), and so then we need to
+pay a lot to duplicate storage and replication work just to maintain the
+independent availability of each index.
+
+# Working Assumptions
+
+The bulk of this proposal is founded on the idea that at some time in the
+future, we will be able to dramatically reduce the clock uncertainty that we
+feel comfortable running with, either across the board or at least within
+CockroachCloud.
+
+Wide area network (WAN) round-trip time (RTT) latencies can be as large as 200ms
+in the global deployment scenarios that we are interested in (ref:
+[AWS](https://docs.aviatrix.com/_images/inter_region_latency.png),
+[Azure](https://docs.aviatrix.com/_images/arm_inter_region_latency.png),
+[GCP](https://docs.aviatrix.com/_images/gcp_inter_region_latency.png)). The
+current default clock uncertainty offset is 500ms. This proposal explores the
+half of the design space in which clock uncertainty bounds drop below WAN RTT
+time. Within this regime, it becomes cheaper to establish causality between
+events using clocks and waiting than by communicating with distant nodes.
+
+As such, the uncertainty bounds necessary for this proposal to make sense are
+much lower than what we currently use, but are also well within the realm of
+possibility for software-only solutions. Clouds are continuing to invest into
+the software-level time synchronization services that they provide. For
+instance, Amazon released a [Time Sync
+Service](https://aws.amazon.com/about-aws/whats-new/2017/11/introducing-the-amazon-time-sync-service/)
+in 2017 that strives to provide reliable time references for EC2 instances.
+Similarly, software-level time synchronization has seen interest from the
+academic community in recent years [4], with research even leading to
+[productionized service offerings](https://www.ticktocknetworks.com/) in some
+cases. Lastly, users are already running CockroachDB with battle-tested time
+synchronization services like [Kronos](https://github.com/rubrikinc/kronos)
+which we may eventually explore using or extending.
```

Comment:
It doesn't look like it. The README says:
> Kronos does not attempt to measure RTT variance. It is not optimized for cross data-center deployments yet

I don't have a grasp on how much work supporting Kronos over WAN would be or whether doing so is feasible.

Reply:
It turns out that WAN support is not as difficult as it might first seem. If L is the minimum latency between two locations A and B, then clocks in those two locations can drift by as much as L with respect to one another, without compromising strict serializability properties. This is because it takes a minimum time L for a transaction finished at location A to cause a transaction started in location B. To illustrate:

Let L = 100ms
Let Uncertainty = 1ms
Let B's clock = A's clock minus 100ms

For transaction A to cause transaction B, it takes a minimum of 100ms after the completion of A. By that time, B's clock has caught up with the commit timestamp of transaction A, and so B will always be ordered after A (subject to the uncertainty interval of 1ms, that is).

Reply:
> For transaction A to cause transaction B, it takes a minimum of 100ms after the completion of A

that's assuming you have a guaranteed lower bound on latency.
If for example you start with a general bad WAN link, and the operator upgrades to something better which shaves off 30%  from L, then we get our uncertainty back.

I think that if you want to play with minimum latencies we need a way to verify that the assumption is not violated. it seems hard to *enforce* a minimum latency either (across a WAN).

Reply:
L is the minimum latency between two locations. That can be measured by having the longitude/latitude of those two locations, and then using the speed of light around the curved surface of the earth, in as direct a path as possible. That is ~20ms between NYC and Paris, for example. Still quite a bit of extra buffer to work with.

Reply:
Also, we don't _enforce_ our uncertainty interval today, at least not in any kind of guaranteed way. We may detect two clocks are too far out of sync. But we may not. Our goal is to _reduce the likelihood_ of violations to a level that is so low that there is no practical reason to worry about it. I think if we go down this path, we'd want to set up a long-running, large-scale test of AWS/GCP clocks on CC machines. We can experimentally determine how well clocks stay sync'd, over a long period of time and under many conditions.

IMO, we'd want to support low uncertainty intervals only in AWS/GCP - either in CC clusters or in clusters that are self-hosted in AWS/GCP. Supporting this in customer DC's should not be a goal, unless it's part of a paid consulting relationship.